### ISSUE TITLE
pubsub.run_in_thread : extra loop requested, for reuse purposes of the pubsub thread when switch to another redis server is needed

### ISSUE THREAD

**Issue by tw-bert** [andymccurdy/redis-py#822]
Hi,

We have a requirement to change running processes to another redis instance on another server on the fly (without restarting a process, and without introducing new threads). This happens within a redis pubsub event. After the event handler has ended, the event handler must be registered to the pubsub channel of the new instance.
 
We implemented this (works great), but it needed a small hotfix on redis-py. Below is the change (not using git atm -hg-, but I could supply a github pull request in a couple of days if requested).

Please change (`client.py`):
```
        pubsub = self.pubsub
        sleep_time = self.sleep_time
        while pubsub.subscribed:
```
to:
```
        pubsub = None
        while not self.pubsub is None and not pubsub is self.pubsub:
            pubsub = self.pubsub
            sleep_time = self.sleep_time
            while pubsub.subscribed:
```

This enables re-assigning the pubsub object to another one while reusing the same thread (and if desired: reuse the original event handler dictionary). The inner loop is restarted again if the self.pubsub member is set to another object during execution of the event handler.

Kind regards, TW

**Comment by Grokzen** [comment]
@tw-bert Why not make a PR with the changes and propose it that way?

**Comment by andymccurdy** [comment]
In your proposed change, how does the inner loop exit? There's nothing in the code you posted that causes the existing pubsub object to become unsubscribed.

I don't see how this works if `sleep_time=None` unless you're calling `unsubscribe()` from somewhere else.

Given you're constructing a new pubsub object to substitute, why not just do something like:

```python
class PubSubThreadManager(object):
    def __init__(self):
        self._thread = None

    def use_pubsub(self, pubsub):
        if self._thread is not None:
            self._thread.stop()
        self._thread = pubsub.run_in_thread()
```

Instantiate this in your code base as a singleton. At startup or whenever the Redis server changes simply construct a pubsub instance and feed it to the `use_pubsub` method.

**Comment by tw-bert** [comment]
Hi Andy, it works with `sleep_time=None` because the request comes from the outside, incoming in the pubsub socket. So you stay inside the loop, but disconnect all and reconnect another, and then `sleep_time=None` will have it's effect in the `select()` of the *new* socket.

I can't use your proposed PubSubThreadManager, because I have no control over the main thread (which isn't python -> embedded python interpreter). And if I would initiate yet another thread to control it, the old thread will never die on OS level. Even `_thread.stop()` does not do that for you. At least not on AIX. The result would be that each switch a thread would be added.

**Comment by andymccurdy** [comment]
Note that the `PubSubWorkerThread.stop` simply unsubscribes from all channels and patterns. It doesn't actually forcibly stop the thread.

But I get that this solution won't work given that you're handling the reconnection in response to a message.

My issue with your proposal above is that it's just not very general purpose. The code doesn't do anything without the associated handler. I'm fine introducing new code, but let's make something others can actually use.

**Comment by tw-bert** [comment]
I agree that it has to be general purpose. The code I suggested makes it possible to change the pubsub object in an event handler. But I agree that my implementation is not intuitive, at least not without comment. Maybe add a method to the client, `swap_pubsub()` ?

I also looked into alternatives like inheriting from `PubSubWorkerThread` and have my own `run_in_thread`, but that way we deviate a lot from the standard redis-py implementation.
